cantusdb/turtle/csv2rdf_by_RDFLib.py

```diff
@@ -0,0 +1,71 @@
+from rdflib import Graph, URIRef, Literal
+from rdflib.namespace import RDF
+import csv
+import validators
+import sys
+
+# helper to get the predicate relation dictionary
+def get_ontology_dict () :
+    
+    # a specific config file for the predicates
+    # TODO: maybe change this part to make a more customized way of getting the predicates configurations
+    ontology_f = open("ontology_dict.csv", 'r')
+    ontology_reader = csv.DictReader(ontology_f)
+    ontology_dict = {}
+    
+    # converting the reader dict to the predicate dictionary
+    for row in ontology_reader:
+        ontology_dict[row[list(row.keys())[0]]] = row[list(row.keys())[1]]
+                
+    return ontology_dict
+    
+# a main operation function
+def convert_csv_to_turtle():
+    # TODO: maybe change this one to accomadate for multiple files
+    if len(sys.argv) != 2:
+        raise ValueError("Invalid number of input filename") 
+    
+    g = Graph()
+
+    csv_file = open(sys.argv[1], 'r', encoding='utf-8')
+    
+    csv_reader = csv.DictReader(csv_file)
+    ontology_dict = get_ontology_dict()
+
+    # Convert each row to Turtle format and add it to the output
+    for row in csv_reader:
+        # the first column as the subject
+        absolute_uri = URIRef(row[list(row.keys())[0]])
+        g.add((absolute_uri, RDF.type, URIRef("http://www.wikidata.org/prop/direct/chant")))
+        
+        # extracting other informations
+        for k in row:
+            # if this key is the first column, skip
+            if k.__eq__(list(row.keys())[0]):
```

Comment:
Why are you using `__eq__`? Just use `==`. And the DictReader gives you the ability to use this as a dict, so getting the first entry in a list of keys completely passes that.

Reply:
The first entry should be the first entry in the dictionary. The first entry in the dictionary should be the key attribute of each entry in the CSV file, it's not the header file. The purpose of this if statement is because the absolute_uri key attribute is already extracted and should be the subject of each triple relation of this entry.

Reply:
`__eq__(self)` is the method that defines how the `==` operator compares two things. Overriding it allows you to specify how you want equality to be computed, and is generally reserved for when you have some complex data type where equality isn't straightforward. 

In this case, though, I think you're just comparing strings? So `k == [whatever]` does the exact same thing.

https://docs.python.org/3/library/operator.html#operator.__eq__

Reply:
Yes, I will change this. I was used to java and had a little confusion.

Reply:
> The first entry in the dictionary should be the key attribute of each entry in the CSV file, it's not the header file

In your CSV file the first row gives your column headers. When you open your CSV file with the DictReader, and do not supply a list of fieldnames, it will interpret the first row as your field names. Thus:

```csv
absolute_url,incipit,genre,genre_id_url,mode_name,mode_id_url,finalis,src_link,cantus_id,indexing_notes,source_description
https://cantusdatabase.org/chant/562633,Ecce nunc palam loqueris et,Antiphon,https://www.wikidata.org/wiki/Q582093,Hypomixolydian mode,https://www.wikidata.org/wiki/Q57976142,G,https://cantusdatabase.org/source/123756,002530,,
```

Becomes: 

```py
{"absolute_url": "https://cantusdatabase.org/chant/562633",
"incipit": "Ecce nunc palam loqueris et",
"genre": "Antiphon",
"genre_id_url": "https://www.wikidata.org/wiki/Q582093",
... etc. }
```

This is done for you so that when you iterate over each row you don't need to access the fields by index, but can simply use the column names in that row: `row["absolute_url"]` rather than `row[list(row.keys())[0]]`.

Reply:
Not in all cases the key attribute is named "absolute_url". It depends on the mapping, but it is always going to be the first entry in the dictionary.

Reply:
That sounds like a recipe for a real headache. :-D